core/src/View/MDSpan/Kokkos_MDSpan_Extents.hpp

```diff
@@ -0,0 +1,156 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#include <Kokkos_Macros.hpp>
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
+static_assert(false,
+              "Including non-public Kokkos header files is not allowed.");
+#else
+KOKKOS_IMPL_WARNING("Including non-public Kokkos header files is not allowed.")
+#endif
+#endif
+
+#ifndef KOKKOS_EXPERIMENTAL_MDSPAN_EXTENTS_HPP
+#define KOKKOS_EXPERIMENTAL_MDSPAN_EXTENTS_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+#ifdef KOKKOS_ENABLE_IMPL_MDSPAN
+
+#include <experimental/mdspan>
+
+namespace Kokkos {
+namespace Impl {
+template <class DataType>
+struct ViewArrayAnalysis;
+
+template <std::size_t... Vals>
+struct ViewDimension;
+
+template <class T, class Dim>
+struct ViewDataType;
+}  // namespace Impl
+
+namespace Experimental::Impl {
+
+/*
+ * A few things to note --
+ * - mdspan allows for 0-rank extents similarly to View, so we don't need
+ * special handling of this case
+ * - View dynamic dimensions must be appear before static dimensions. This isn't
+ * a requirement in mdspan but won't cause an issue here
+ */
+
+template <std::size_t N>
+struct ExtentFromDimension {
+  static constexpr inline std::size_t value = N;
+};
+
+/*
+ * Kokkos uses a dimension of '0' to denote a dynamic dimension.
+ */
+template <>
+struct ExtentFromDimension<std::size_t{0}> {
+  static constexpr inline std::size_t value = std::experimental::dynamic_extent;
+};
+
+template <std::size_t N>
+struct DimensionFromExtent {
+  static constexpr inline std::size_t value = N;
+};
+
+template <>
+struct DimensionFromExtent<std::experimental::dynamic_extent> {
+  static constexpr inline std::size_t value = std::size_t{0};
+};
+
+template <class SizeType, class Dimension, class Indices>
+struct ExtentsFromDimension;
+
+template <class SizeType, class Dimension, std::size_t... Indices>
+struct ExtentsFromDimension<SizeType, Dimension,
+                            std::index_sequence<Indices...>> {
+  using dimension_type = Dimension;
+  using type           = std::experimental::extents<
+      SizeType,
+      ExtentFromDimension<Dimension::static_extent(Indices)>::value...>;
+};
+
+template <class Extents, class Indices>
+struct DimensionsFromExtent;
+
+template <class Extents, std::size_t... Indices>
+struct DimensionsFromExtent<Extents, std::index_sequence<Indices...>> {
+  using extents_type = Extents;
+  using type         = ::Kokkos::Impl::ViewDimension<
+      DimensionFromExtent<extents_type::static_extent(Indices)>::value...>;
+};
+
+template <class DataType>
+struct ExtentsFromDataType {
+  using array_analysis = ::Kokkos::Impl::ViewArrayAnalysis<DataType>;
+  using size_type      = std::size_t;  // Mirrors Kokkos::View's size type
```

Comment:
Why wouldn't we go ahead and stop pretending that it is configurable in `ExtentsFromDimension` and not have it as a template parameter.  Even if it was we still wouldn't be able to deduce the right thing from just the `DataType`.